3.1 RFC: Cookies

robertpustulka opened the pull request:
Here is some draft code of my proposal of cookies handling - following #6717.

In short - this aims to detatch cookie manipulation from a Controller layer. In the past cookie encryption and complex data handling was possible only through CookieComponent which made it harder to manipulate cookies outside Controllers/Components forcing client classes to depend at least on ComponentRegistry.

It's just a draft, I intend to improve this later. I'm also open to suggestions.

robertpustulka committed:
Cookies refactoring draft.

jadb commented:
Editing at the same time doesn't help, lol... Re-adding enhancement to the labels.

markstory commented on src/Network/Cookie/RequestCookieJar.php:
```diff
@@ -0,0 +1,19 @@
+
+    public function __construct()
+    {
+        foreach ($_COOKIE as $name => $value) {
```
I would take the cookies as a parameter. Right now there is no way for a fake request to easily pass cookies in. In general I've been trying to herd superglobal references into as few places as possible.

markstory commented on src/Network/Cookie/RequestCookieJar.php:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    public function get($name) {
+        return $this->_cookies[$name];
```
Should you have set() and remove() as well?

robertpustulka replied:
These are cookies sent with a request, so I don't think they should be modified.

markstory commented on src/Network/Cookie/ResponseCookieJar.php:
```diff
@@ -0,0 +1,43 @@
+        return $cookie;
+    }
+
+    public function add(Cookie $cookie) {
```
add() and create() don't seem that different. Why not have one method that accepts multiple types of arguments.

robertpustulka replied:
OK, fair point.

markstory commented on src/Network/Cookie/ResponseCookieJar.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    public function set() {
+        foreach($this->_cookies as $cookie) {
+            setcookie(
```
I wouldn't do this. Instead the response should use the cookie jar to get the cookie data. The response should be responsible for sending headers not the cookie jar.

robertpustulka replied:
OK. I'll make those jars iterators anyway so there wouldn't be a problem for request to traverse it.

robertpustulka committed:
Updated cookie jars.

robertpustulka committed:
Updated cookie + cookie encryption class.

robertpustulka committed:
Renamed delete to remove.

robertpustulka commented:
I've updated cookie jars. Now they extend base `AbstractCookieJar`, which provides simple getter and iterator. I've also removed hardcoded `$_COOKIES` var and `setcookie()`.

Next I've extended `Cookie` class with getters/setters for cookie "properties" (path, domain, expires etc) as wrappers for config. I've moved cookie encryption logic to `CookieEncryptor` class which (for now) provides static methods for cookie encryption and encoding - this is just a copy of `CookieComponent` code. Not sure if it needs any upgrade.

There may be a problem with decryption of cookies sent with a request. By default cookie isn't aware of whether it is encrypted (and how) or not. For now I came up with keeping default cookie config in `app.php` config and access it via `Configure`, but I'm not sure whether it is a good way to do it...

robertpustulka committed:
Updated docblocks.

robertpustulka committed:
Docblocks + use clauses.

robertpustulka committed:
Added the ability to set a raw cookie data.

robertpustulka committed:
Returtn value of expires() fixed.

markstory commented on src/Network/Cookie/Cookie.php:
```diff
@@ -0,0 +1,256 @@
+        'secure' => false,
+        'salt' => null,
+        'httpOnly' => false,
+        'encryption' => 'aes',
```
Could encryption be a subclass? Would that make decrypting cookies later easier?

markstory replied:
Or perhaps encryption belongs in the jar? That would help centralize it and prevent cookies from needing to know too much.

robertpustulka replied:
@markstory even if encryption was a subclass request wouldn't know that while setting cookies. That information should be somehow pre-configured (like in CookieComponent) or kept inside cookie value, but that doesn't feel secure...

robertpustulka replied:
I wonder whether cookie encryption should be opt-in or opt-out. That could help with implementation.

ADmad replied:
Currently it's opt-out in CookieComponent and should stay same.

robertpustulka replied:
But non-existent in Request/Response... So we decide that all cookies should be encrypted wit aes. Is it reasonable to leave rijndael?

ADmad replied:
No, for backwards compatibility it should be same as cookiecomponent.

robertpustulka replied:
OK, but there still is a problem how to detect that cookie sent with previous response wasn't encrypted. It needs to be configured that way somewhere so in succeding requests/responses cookies know what they are.

robertpustulka replied:
In laravel framework they do encryption in middleware. So its an external process - cookies themselves are not aware of encryption and that may be the right approach.

So let's say we move cookies encryption to another class and handle it from there at some point of app runtime. There should be some way of telling this encrypter what to do with cookies -should they be encrypted and how as CookieComponent allows cookies to be configured that way on the name basis.

ADmad replied:
Yeah since we can configure encryption per cookie we od need a way for each cookie to tell the encrypter whether it needs to be encrpted / decrypted using appropriate scheme.

markstory commented on src/Network/Cookie/Cookie.php:
```diff
@@ -0,0 +1,256 @@
+            $config = ['name' => $config];
+        }
+
+        if (empty($config['name'])) {
```
empty() will explode on '0'.

markstory commented on src/Network/Cookie/Cookie.php:
```diff
@@ -0,0 +1,256 @@
+            return $value;
+        }
+
+        $expires = new Time($value);
```
Why not return the time instance?

robertpustulka replied:
True, request could handle expire format for setcookie()

markstory commented on src/Network/Cookie/RequestCookieJar.php:
```diff
@@ -0,0 +1,24 @@
+    public function __construct(array $cookies)
+    {
+        foreach ($cookies as $name => $cookieData) {
+            $cookieData += (array)Configure::read("Cookies.$name");
```
Why?

robertpustulka replied:
It's a temporary resolution for cookie encryption setup.

markstory commented on src/Network/Cookie/RequestCookieJar.php:
```diff
@@ -0,0 +1,24 @@
+            $cookieData['name'] = $name;
+
+            $cookie = new $this->_cookieClassName($cookieData);
+            $this->_cookies[$cookie->name()] = $cookie;
```
You could/should just reuse the add() method here.

robertpustulka replied:
@markstory RequestCookieJar doesn't have add() method. I;m wondering whether it should have it and should we be able to add cookies to te request on runtime.

markstory replied:
Probably shouldn't be able to add cookies, at least not outside of test cases.

lorenzo commented on src/Network/Cookie/AbstractCookieJar.php:
```diff
@@ -0,0 +1,42 @@
+     */
+    public function getIterator()
+    {
+        return $this->_cookies;
```
An array is not an iterator

robertpustulka replied:
Right, what about?
```php
    public function getIterator()
    {
        return new ArrayIterator($this->_cookies);
    }
```

lorenzo replied:
That would be the way to go :)

lorenzo commented on src/Network/Cookie/Cookie.php:
```diff
@@ -0,0 +1,256 @@
+     *
+     * @return \Cake\Network\Cookie\Cookie
+     */
+    public function forget()
```
I would name this `invalidate`

robertpustulka committed:
Fix iterator

robertpustulka committed:
Cookie updates

robertpustulka committed:
Rename forget to invalidate

robertpustulka committed:
Made cookie encryption-agnostic

robertpustulka committed:
CookieEncryptor is now CookieEncrypter and is no longer "static class"

robertpustulka committed:
Moved cookies encryption into cookie jars.
Made lazy cookie loading and sending.

robertpustulka committed:
Removed unused config keys

robertpustulka commented:
Here is another attempt to solve cookie encryption.

I've made cookies object creation "lazy". Raw cookie data is loaded into a seperate array and every` get()` operation creates an object with optional encryption setting. That allows you to decrypt cookies on demand. 

That works the other way. If you want to send cookie unencrypted you can call `send()` with optional encryption parameter. That converts cookie object back to an array, performs encryption and moves data to seperate array removing cookie from an object array.

Just before the response is sent back to browser `Response` can retrieve prepared cookies using `raw()` method. All remaining cookie objects are transformed into arrays and returned with that method.

I'm not sure about `raw` and `send` method names...

robertpustulka committed:
Don't enforce cookie value to be an array

markstory commented on src/Network/Cookie/ResponseCookieJar.php:
```diff
@@ -0,0 +1,91 @@
+     * @param bool|string|array $encryption
+     * @return void
+     */
+    public function send($name, $encryption = null)
```
This method doesn't actually send anything.

robertpustulka replied:
Yeah, I'm looking for another name. Any suggestions?

robertpustulka replied:
Maybe `queue`?

robertpustulka committed:
Fixed access methods.

robertpustulka committed:
Made a CS a bit happier

robertpustulka committed:
Rename method "send" to "queue".

robertpustulka committed:
Added "has" and "unqueue" methods.

robertpustulka committed:
Unqueue fixed to conform with add method API.

robertpustulka committed:
Unset unqueued raw cookie data.

robertpustulka committed:
DRY the code

robertpustulka committed:
Request cookies should be immutable. So no cookie objects are cached inside request jar, and every get() method returns a fresh Cookie instance made out of the same raw cookie data.

markstory commented on src/Network/Cookie/ResponseCookieJar.php:
```diff
@@ -0,0 +1,109 @@
+            unset($this->_rawCookies[$name]);
+
+            return $this->add($cookie);
+        }
```
What if instead of the queue/unqueue methods there was a method that accepted a response class and set the cookie data into the response? Alternatively a method to get the cookies out of the jar in a format that the response could understand?

Right now I don't understand the value of these methods.

rchavik replied:
How about `dequeue` for the name?

robertpustulka replied:
@markstory these methods will be used very rarely. They provide some way of overwriting default encryption for cookies, that need that for some reason. By default you just get cookies from a request, do what you want to do with them and add to the response. But you may want some cookies to have encryption disabled. You get these cookies from the request (`get()` with `$encryption = false`) and also mark them as ready for the response with no encryption enabled in that case (using `queue` method with optional argument).

robertpustulka commented:
Please let me know if my idea behind cookie queueing and encryption/decryption is clear. I'm also open to any suggestions on how to do this better.

Schlaefer commented:
I would really like to see a nice object based Cookie system and develop this PR further. Ideas:

Jars don't manipulate or manage raw cookie-data or encryption, they are a collection of Cookie objects. The Cookie itself handles (from/to) raw and encryption internally.

RequestJar accepts raw `$_COOKIE` in constructor and creates its Cookies. You can't manipulate the collection afterwards. RequestJar returns only copies of its Cookies.

ResponseJar allows to add or remove Cookies.

Putting it together: RequestJar and ResponseJar are appended as public `cookies` property to Response and Request. So in the future:

```php
$cookie = $controller->request->cookies->read('foo');
$bar = $cookie->get('bar');
…
$cookie->set('bar', 'baz');
$controller->response->cookies->write($cookie);

$cookie = new Cake\Network\Cookie\Cookie('zap' [, $options]);
$cookie->set('foo', 'bar');
$controller->response->cookies->write($cookie);
```

Response can then iterate over its Jar Cookies when creating the response.

CookieComponent switches internally to the classes above and stays around for BC. `Request::cookies()` and `Response::cookies()` are marked deprecated. All of those can then be removed on a breaking release.

robertpustulka commented:
@Schlaefer but should cookie object know whether it's encrypted or not? You need to tell it somehow if and how it is encrypted prior to reading it's value.

The rest of your suggestions I agree with and they are more or less already implemented in this PR. 

The only problem is this encryption scope. We need some way of decrypting cookies before we manipulate them and encrypting them afterwards. My proposition above is to tell `RequestJar` how to decrypt the cookie while getting it from a request (`RequestJar::get`) and to tell `ResponseJar` how to encrypt it before it is sent with a response (`ResponseJar::queue`). In other words: *unwrap* cookie from a response, manipulate it, add it to the response jar and *wrap* it just before it is sent back to browser.

Schlaefer commented:
> The only problem is this encryption scope.

How's about using the encryption setter before reading values:

```php
$cookie = $controller->request->cookies->read('foo')
    ->encryption('plain');
$bar = $cookie->get('bar');
```

Otherwise every cookie is assumed as AES (default Cookie property).

robertpustulka commented:
@Schlaefer this solution would require some way of disabling cookie encryption setter after the value has been decrypted and manipulated.

Schlaefer commented:
Thinking on it further: ultimately you want a specific cookie-class. Retrieving always the same cookie class from the Request doesn't make sense. So no RequestJar, but something like

```php
$cookie = new Cookie('foo');
$cookie->setData($controller->request);

$cookie = new AuthCookie();
$cookie->setData($controller->request);

```

With that it's always clear which encryption to apply ($cookie->encryption() when 'setData' is called).

robertpustulka commented:
How that cookie class would react to changing its encryption method after `setData` call or multiple `setData` calls?

Schlaefer commented:
As I said, the Cookie handles the `setData()` (shitty name I know) with the currently set encryption. You could also provide the encryption with an additional argument to that method, but let's wait until someone requests that feature.

There's only one request per cycle and that data never changes, so one would assume a Cookie is only initialized once from request data.

Is changing the encryption a real world issue? Usually you decide on an encryption and that's how you save and read it from then on. If the reading fails, e.g. wrong format, you have to handle that manually anyway.

robertpustulka commented:
OK, that seems reasonable.

But I'd like to have some "factory" for request cookies. Having to do this:
```php
$cookie = (new Cookie('foo'))->encryption(false)->setData($this->request)
``` 
every time could become tedious.

I'd like to see something like: `$this->request->cookies->create('foo', ['encryption'=>false])` also.
This would still leave the door open for handling descendant cookie class objects manually.

markstory commented:
To me encryption is a feature of the jar, not the individual cookies. The jar uses an encrypter that knows which cookies are encrypted. This information would need to be decided apriori with config.

```php
// Used by itself
$encrypter = new CookieEncrypter('aes_sekret', ['secret_cookie', 'other_secret_cookie']);
$jar = new RequestJar($_COOKIE, $encrypter);
$jar->get('secret_cookie')->value();

$jar = new ResponseJar([], $encrypter);
$jar->set('secret_cookie', $cookieData);
```

In general we should also be thinking about how a new cookie implementation could be compatible with PSR-7. From reading the specs injecting cookie data into requests as part of middleware would be viable and I don't think we've discussed anything here that would make that hard.

robertpustulka commented:
What about configuring encrypter in a middleware = dispatcher filter? I think creating and attaching jars to request/response is a job for a filter. Maybe encrypter could also be attached that way? This could ensure jars have the same encrypter instance with the same config. Of course there should be some way of configuring encrypter at runtime to provide BC for `CookieComponent`.

Schlaefer commented:
I still strongly believe that encryption is a cookie property and will provide a cleaner solution in the long run.

Otherwise "encryption" would be a item property depended on the collection it is in, that doesn't feel right. Also you couldn't append one RequestJar to the Request, because you need different jars for different encryption types (yes, I actually use that).

Having a RequestJar feels™ superfluous at the moment. I like the idea of Cookie-classes being preconfigured "entities" you "patch" with "request-data".  Have to sleep about it. Maybe I wipe out some code over the weekend.

markstory commented:
@Schlaefer I don't like that every cookie needs the AES key though.

robertpustulka commented:
How about this?
```php
class CookieEncryptionFilter extends DispatcherFilter
{

    public function beforeDispatch(Event $event)
    {
        $request = $event->data['request'];
        $this->_encrypter->decrypt($request->cookies);
    }

    public function afterDispatch(Event $event)
    {
        $response = $event->data['response'];
        $this->_encrypter->encrypt($response->cookies);
    }
}
```

That will make cookie encryption independent of jars and keep the code cleaner.

robertpustulka committed:
Updated encrypter to allow for per-cookie configuration

robertpustulka committed:
Jars updated

robertpustulka committed:
Added cookie filter

robertpustulka commented:
I've updated the code with an encrypter configurable for each cookie.

For example in a controller we can do this for custom encryption:
```php
$this->request->encrypter()->config('foo', ['encryption' => 'rijndael', 'key' => 'XYZ']);
$cookie = $this->request->get('foo');
$cookie->write('bar');
$this->response->set($cookie);
```

robertpustulka committed:
Made encrypter required in jar constructors

robertpustulka committed:
Removed _rawCookies property.

Schlaefer commented:
Ok, [here's my take on the issue][1] after fiddling on an implementation. Not polished (as nobody fancies encryption as Cookie property anyway), but working and you get the idea.

No Jars. The CookieComponent uses the Cookie class internally. No API changes, no changes to Routes and minor internal changes to Response. Everything else would require a breaking release, to much code changes or had a difficult to use interface.

With BC in place, Cookie classes could be exposed via new methods in CookieComoponent. A getter would produce a standard Cookie class with the current CookieComponent settings. Another method would allow to register individual Cookie classes for a particular cookie-name:


```php
// in app bootstrap
$cookie = new Cookie('foo')->secure(true)->expires('+1 week');
$controller->Cookie->register($cookie);
// or you have a already configured Cookie subclass
$controller->Cookie->register(new AuthCookie());

// … later …
$cookie = $controller->Cookie->get('foo');
$cookie = $controller->Cookie->get('auth');
$cookie->get('username');
```

[1]: https://github.com/cakephp/cakephp/commit/84a5fe2eb38ec98637da28cb3bd9767d4007bc81

robertpustulka commented:
The aim of this PR is to decouple cookies from controller/component.
Ideally I'd see CookieComponent gone - too many unwanted dependencies.

-- 
Pozdrawiam,
Robert Pustułka

ADmad commented:
> Ideally I'd see CookieComponent gone - too many unwanted dependencies.

@robertpustulka 3.x needs to be backwards compatible so can't do that till 4.0.

Schlaefer commented:
@robertpustulka Sorry, didn't want to interrupt, just present ideas.

That said I think it's important to note that you can't remove the CookieComponent in a non-breaking release. Also you don't touch the CookieComponent in your PR nor outline (or provide code) how the user is supposed to interact with cookies after CookieComponent is gone.

robertpustulka commented:
@ADmad, I mean - depracate CookieComponent in 3.1 and remove it in 4.0.

-- 
Pozdrawiam,
Robert Pustułka

robertpustulka commented:
@Schlaefer I'm waiting with CookieComponent until Cookie API becomes clear more or less. I still have CookieComponent BC back in my head and I think it should be quite easy to upgrade the component.

I'm not against any new ideas, it could be great if we could make cookie API flexible, easy to configure and independent of a controller layer.

I think your porposal isn't much far from what has emerged from this PR. You could also extend base Cookie class with your own preconfigured subclass and set objects to the response with `$this->response->cookies->set(new YourCookie())`. The only difference is cookie encryption which I believe cookie objects shouldn't be aware of.

Schlaefer commented:
> On 25.06.2015, at 17:59, Robert Pustułka <notifications@github.com> wrote:
> 
> I still have CookieComponent BC back in my head and I think it should be quite easy to upgrade the component.

OK, I wasn't sure. At the moment those PR classes feel like they hang in the air a little bit. And CookieComponent BC has a few quirks too, as I discovered today. ;)

robertpustulka commented:
@Schlaefer, yeah they do - or they did as I've updated PR today with a
dispatcher filter that presents my idea of glueing them together.

AFAIR CookieComponent treats cookies as a one container for cookie data
with the optional ability to configure each cookie seperately. This may
become troublesome to port I'm afraid :/. The code seems complex but
the idea may be simple ;)


-- 
Pozdrawiam,
Robert Pustułka

sitedyno commented:
@robertpustulka 
> I've moved cookie encryption logic to CookieEncryptor class which (for now) provides static methods  for cookie encryption and encoding - this is just a copy of CookieComponent code. Not sure if it needs any upgrade.

This seems like a good time to introduce another (better) encryption/decryption method as described here: https://paragonie.com/blog/2015/05/using-encryption-and-authentication-correctly I'm pretty sure Scott would be willing to do a code review if asked.

Since requiring libsodium is probably a hard sell; note that he does mention a pure PHP alternative (requires defuse/php-encryption >=5.4): https://gist.github.com/wopot/94e33bdd1d7faaaa56e2

markstory commented:
@Schlaefer I'm not a big fan of the aes key being in each of the cookies. To me cookie objects should act like value objects and not be overly smart. I'm also don't think that a CookieJar is a suitable place for encryption to happen.

Having a dispatcher filter that decrypts cookies is another way around the Jar concept, and provides similar functionality to what we have. While I know the concern around backwards compatibility has come up a few times (which is great, because it is very important). We can also build two parallel implementations of cookies. If the goal is to deprecate the component, and have a more flexible decoupled system, we can do that and leave the component as it stands (for the most part). I think @Schlaefer's prototype illustrates this pretty well.

Schlaefer commented:
> On 26.06.2015, at 03:50, Mark Story <notifications@github.com> wrote:

> To me cookie objects should act like value objects and not be overly smart. … Having a dispatcher filter that decrypts cookies is another way around the Jar concept

The encryption logic could be moved and handled in a dispatcher filter. But if the transport properties are handled there you can't lazy evaluate cookies anymore?

In that case would it be feasible to avoid manipulating cookie transport-properties on the cookie itself? Separate data and configuration completely. Instead define cookie configurations (predefine in bootstrap) and let the filter dispatcher act on those configurations. The cookie in the app is just a dump value object getter/setter (or everything with that interface).

robertpustulka commented:
@markstory having decryption being done in a filter leaves configuration to be done during bootstrap before a controller is initialized. `CookieComponent` encryption is configured in a component, so - if I'm correct - after `beforeDispatch` is fired.

@Schlaefer I wouldn't go that far. Having the ability to configure expire, path, domain etc at runtime is a good thing to have. That would leave only encryption to be configured during bootstrap - that is as another config key, where we could also configure enryption for individual cookies.

Maybe we could go that way:
- have encryption done in a filter and configured in app.php
- attach jar(s) in another filter
- force cookie class to implement some `CookieInterface`
- have another `CookieInterface` implementation only for `CookieComponent` with encryption being done in cookies (because CookieComponent have encryption configured in its config that is after cookies have been decrypted in beforeDispatch)

Incoming cookies don't have any properties apart from a name and value. These don't need to be objects and could be attached to a response as an array (with values decrypted in a filter) and - as @Schlaefer suggested - used to feed cookie objects when nessesary.

markstory commented:
@robertpustulka Correct, which cookies were encrypted would need to be decided ahead of time, before the controller got involved. I don't see this as a bad thing necessarily, as its very unlikely that one cookie name will be encrypted in some places, but not others.

@Schlaefer The only thing I'm proposing moving out to the filter is encryption, all the other aspects of cookies could remain mutable in the controller actions.

robertpustulka committed:
Added CookieInterface

robertpustulka committed:
Moved encrypter out of jars

robertpustulka committed:
Added ArrayAccess to cookie jars.

robertpustulka committed:
Moved encryption to another filter.

robertpustulka committed:
Minor CS fixes

robertpustulka commented:
I've pushed a few commits.

Encryption is now done in a `CookieEncryptionFilter` and Cookies implement CookieInterface to allow for custom cookie implementations (encryption being done in a cookie - here you go).

I'm not sure if jars should be injected in a `CookieFilter` - I left that temporarily. I think Request/Response classes could create their jars.

I'm also wondering if there's a need for a RequestJar anymore. Request cookies are just a name and a value, no objects nessessary. Cookie objects could be fed with request data on demand, for example:

```php
$cookie = new Cookie(['name' => 'foo']); 
$cookie->write($response->cookie('foo')); 

//or something like @Schlaefer suggested:
$cookie->feed($response); //cookie could automatically find data in request and write data
...
$response->cookies->set($cookie);

//another way:
$response->cookies->set('foo', $request->cookie('foo'));
$cookie = $response->cookies->get('foo');
...
```

Schlaefer commented:
> I'm not sure if jars should be injected in a CookieFilter - I left that temporarily. I think Request/Response classes could create their jars.

+1 One filter should be able to handle the cookie encryption. 

> I'm also wondering if there's a need for a RequestJar anymore. Request cookies are just a name and a value, no objects nessessary. 

+1

> (encryption being done in a cookie - here you go)

Let's not go there again. I have the feeling that train left the station. :wink: But if another approach is actually preferred the question of when and where to set the encryption configuration for a particular cookie should imho be addressed, because it may affect decisions as those above.

ADmad commented on src/Routing/Filter/CookieEncryptionFilter.php:
```diff
@@ -0,0 +1,100 @@
+    public function beforeDispatch(Event $event)
+    {
+        $request = $event->data['request'];
+        $this->_decryptCookies($request);
```
This means in each request all cookies will be decrypted even if you don't actually need to read them isn't it?  That seems wasteful and unnecessary overhead.

robertpustulka replied:
@ADmad I'm afraid that's the case with encryption being made in a filter. Check previous commits where I proposed lazy cookie decryption. Different solution has been mentioned or implemented here. It's just a matter of deciding which is the best one.

robertpustulka commented:
I guess I'll hold this PR until #6960 is decided. 
I'll keep this open as a reminder and then retarget it for 3.2 in a new PR.

robertpustulka commented:
The issue has been closed but what about the idea?
